### ISSUE TITLE
Alice2 - uniq tag resolves to empty list. also crashes when asked "how is A like B"

### ISSUE THREAD

**Issue by mrezzamoradi** [keiffster/program-y#94]
Hello @keiffster and all, thanks for the job.

I wasn't sure if it is okey to create any issues about Alice2, since they were included only for completeness, so I brought two problems in one issue. sorry for that.

The problems
--------------

I started alice2 console client (alice2.sh) just to test some questions. I found these two problems:

1. When asking "**sound dog?**", she replies with "**dog says .**"
2. When asking "**how is cat like lion?**", the bot crashes after about a minute.

Extra info for problem 1
------------------------
This is the category in _animal.aiml_ that must match:
```
<category>
        <pattern>SOUND
            <set>animal</set>
        </pattern>
        <template>
            <star/>
            says
            <uniq>
                <subj>
                    <star/>
                </subj>
                <pred>sound</pred>
                <obj>?sound</obj>
            </uniq>
            .
        </template>
</category>
```
and this is the related part of _alice2.log_:
```
2017-10-21 17:01:18,889  root       DEBUG     Matches...
2017-10-21 17:01:18,889  root       DEBUG       1 - Match=(Word) Node=(WORD [SOUND]) Matched=(SOUND)
2017-10-21 17:01:18,889  root       DEBUG       2 - Match=(Word) Node=(SET name=[ANIMAL]) Matched=(dog)
2017-10-21 17:01:18,890  root       DEBUG       3 - Match=(Topic) Node=(ONEORMORE [*]) Matched=(*)
2017-10-21 17:01:18,890  root       DEBUG       4 - Match=(That) Node=(ONEORMORE [*]) Matched=(*)
2017-10-21 17:01:18,890  root       DEBUG       T - PTEMPLATE [P(0)^(0)#(0)C(0)_(0)*(0)To(0)Th(0)Te(1)] 
2017-10-21 17:01:18,890  root       DEBUG   AIML Parser evaluating template [PTEMPLATE [P(0)^(0)#(0)C(0)_(0)*(0)To(0)Th(0)Te(1)] ]
2017-10-21 17:01:18,890  root       INFO    Retrieving conversation for client Console
2017-10-21 17:01:18,890  root       DEBUG   Star Node [STAR] resolved to [dog]
2017-10-21 17:01:18,890  root       DEBUG   [[WORD]says] resolved to [says]
2017-10-21 17:01:18,890  root       INFO    Retrieving conversation for client Console
2017-10-21 17:01:18,890  root       DEBUG   Star Node [STAR] resolved to [dog]
2017-10-21 17:01:18,890  root       DEBUG   [[NODE]] resolved to [dog]
2017-10-21 17:01:18,891  root       DEBUG   [[WORD]sound] resolved to [sound]
2017-10-21 17:01:18,891  root       DEBUG   [[NODE]] resolved to [sound]
2017-10-21 17:01:18,891  root       DEBUG   [[WORD]?sound] resolved to [?sound]
2017-10-21 17:01:18,891  root       DEBUG   [[NODE]] resolved to [?sound]
2017-10-21 17:01:18,891  root       DEBUG   [UNIQ] resolved to []
2017-10-21 17:01:18,891  root       DEBUG   [[WORD].] resolved to [.]
2017-10-21 17:01:18,891  root       DEBUG   [[NODE]] resolved to [dog says  .]
2017-10-21 17:01:18,891  root       DEBUG   Raw Response (Console): dog says  .
2017-10-21 17:01:18,910  root       DEBUG   Denormalising input from [dog says  .] to [dog says .]
2017-10-21 17:01:18,910  root       DEBUG   Formatting punctuation...
```
There's an entity in _config/triples.txt_ which is `dog:sound:woof` and the `<uniq>` tag is supposed to return it, but instead it returns an empty list.

Extra info for problem 2
------------------------
After asking the question, one of my CPU cores went up to 100% usage. This is the category in _triple.aiml_ that must match:
```
<category>
        <pattern>HOW IS * LIKE *</pattern>
        <template>
            <think>
                [A lot fo lines]
</category>
```
It seems that the program goes into an infinite loop:
```
  File "/home/program-y/src/programy/parser/template/nodes/condition.py", line 425 in resolve
  File "/home/program-y/src/programy/parser/template/nodes/condition.py", line 471 in resolve_type2_condition
  File "/home/program-y/src/programy/parser/template/nodes/condition.py", line 425 in resolve
  File "/home/program-y/src/programy/parser/template/nodes/condition.py", line 471 in resolve_type2_condition
  File "/home/program-y/src/programy/parser/template/nodes/condition.py", line 425 in resolve
  File "/home/program-y/src/programy/parser/template/nodes/condition.py", line 471 in resolve_type2_condition
  ...
alice2.sh: line 7: 20109 Aborted                 (core dumped) python3 ../../src/programy/clients/console.py --config ./config.yaml --cformat yaml --logging ./logging.yaml
```

Many thanks

**Comment by keiffster** [comment]
In terms of problem 1 I think I’ve found an issue with case matching, fix shouldn’t take long

As for problem 2, in Alice this typically points to a broken grammar, I’ll try and track it down but Alice 2 originally is full of bugs an incomplete grammars

K

**Comment by keiffster** [comment]
So I’m now deep into this issue and once again a big thanks you to the pandorabot team and their total lack of documentation around rdf features. This has uncovered yet another use case that needs to be reverse engineered out

Fix coming ....

**Comment by mrezzamoradi** [comment]
The second problem is not really important. I also checked [AliceBot](http://sheepridge.pandorabots.com/pandora/talk?botid=b69b8d517e345aba&skin=custom_iframe) for the same type of question and it always returns "Sounds fine as far as I know". Seems they have not really solved the problem themselves.

Many many thanks for following up the issue, the rdf thing can come handy in expanding the bot. I would have tried to dig it myself, but I am currently a little busy on using the bot for an RTL language.

By the way that would be nice if you add rtl comma, semicolon and question mark `، ؟ ؛` to the `STRIP_ALL_PUNCTUATION` (it really didn't worth a pull request)

**Comment by keiffster** [comment]
uniq issue now fixed in the development branch and will be pushed to master over the next couple days

**Comment by keiffster** [comment]
Released in v1.2.0 and beyond
